### PR TITLE
support active/focused view or panel `when` clause context

### LINKED ISSUE #4104: support active/focused view or panel when closures
See https://code.visualstudio.com/docs/getstarted/keybindings#_activefocused-view-or-panel-when-clause-context

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

- fix #4104: support active/focused view or panel `when` clause context

See https://code.visualstudio.com/docs/getstarted/keybindings#_activefocused-view-or-panel-when-clause-context for requirements. **Important**: in this context _active_ means _visible_, not _focused_ as usual. 

In VS Code there are viewlets, panels and views:
  - viewlet is a tab in the left area (explorer, scm, search, debug, etc.)
  - panel is a tab in the bottom area (terminals, markers, output, debug console, etc.)
  - view is a part in the viewlet, e.g. files and outline are views in the explorer viewlet

**Important difference** to Theia that VS Code does not allow to move individual viewlets or panels to other areas. It means that there is always only one visible viewlet or panel in VS Code, but in Theia it can be that several viewlets and panels widgets are visible. **This PR implements the difference by setting context keys to last focused visible viewlet or panel widget.**

#### How to test

- install https://github.com/akosyakov/vscode-view-when-sample/blob/master/vscode-view-when-sample-0.0.1.vsix
- to test `activeViewlet` and `activePanel`: open quick command palette and type `Test` you should see commands visible in the current context https://github.com/akosyakov/vscode-view-when-sample/blob/dd61713e5124d10e744d02d0694f177b3d46af3f/package.json#L69-L109
- to test `focusedView`: focus contributed `nodeDependencies` view in the explorer tab and type `ctrlcmd+f1`, it should trigger a notification bound to this keybinding when this view is focused: https://github.com/akosyakov/vscode-view-when-sample/blob/dd61713e5124d10e744d02d0694f177b3d46af3f/package.json#L112-L118

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by akosyakov** [comment]
- `repl` is the debug console, you can open it from the debug view
- `search panel` is then search view is located not in the side areas, but in main or bottom

File: packages/core/src/browser/shell/application-shell.ts [hunk 1]
```diff
@@ -205,10 +224,17 @@ export class ApplicationShell extends Widget {
         this.mainPanel = this.createMainPanel();
         this.topPanel = this.createTopPanel();
         this.bottomPanel = this.createBottomPanel();
+
         this.leftPanelHandler = sidePanelHandlerFactory();
         this.leftPanelHandler.create('left', this.options.leftPanel);
+        this.leftPanelHandler.dockPanel.widgetAdded.connect((_, widget) => this.fireDidAddWidget(widget));
+        this.leftPanelHandler.dockPanel.widgetRemoved.connect((_, widget) => this.fireDidAddWidget(widget));
```
Comment (akosyakov) [inline packages/core/src/browser/shell/application-shell.ts]: `fireDidRemoveWidget`, the same in all other places

**Comment by vince-fugnitto** [comment]
> * `repl` is the debug console, you can open it from the debug view
> * `search panel` is then search view is located not in the side areas, but in main or bottom

I quickly tested, both cases work as well :)

File: packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts [hunk 2]
```diff
@@ -478,4 +505,86 @@ export class PluginViewRegistry implements FrontendApplicationContribution {
...
+                })
+            );
+            if (view.sideArea !== undefined) {
+                toDispose.pushAll([
+                    this.shell.onDidAddWidget(w => {
+                        if (w === widget) {
+                            this.updateVisibleWidget(widget, view);
+                        }
+                    }),
+                    this.shell.onDidRemoveWidget(w => {
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts]: it does not seem to be necessary, listening to `onDidAddWidget` is enough

**Pull request merged**